src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java

```diff
@@ -188,35 +189,105 @@ public static boolean isStringSafe(@CheckForNull String input, @CheckForNull Env
+                output.add(o);
+            }
+        }
+
+        return (isMutated) ? output : objects; // Throw away copies and use originals wherever possible
+    }
+
+    /** Sanitize a single non-Map, non-List, non-{@link Step}/non-{@link UninstantiatedDescribable} item */
+    @CheckForNull
+    Object sanitizeObjectAndRecordMutation(@CheckForNull Object o, @CheckForNull EnvVars vars) {
+        // Package scoped so we can test it directly
+        Object tempVal = o;
+        if (tempVal instanceof Step) {
+            // Ugly but functional used for legacy syntaxes with metasteps
+            tempVal = ((Step)tempVal).getDescriptor().defineArguments((Step)tempVal);
+        } else if (tempVal instanceof UninstantiatedDescribable) {
+            tempVal = ((UninstantiatedDescribable)tempVal).toMap();
+        }
+
+        if (isOversized(tempVal)) {
+            this.isUnmodifiedBySanitization = false;
+            return NotStoredReason.OVERSIZE_VALUE;
+        }
+
+        Object modded = tempVal;
+        if (modded instanceof Map) {
+            // Recursive sanitization, oh my!
+            modded = sanitizeMapAndRecordMutation((Map)modded, vars);
+        } else if (modded instanceof List) {
+            modded = sanitizeListAndRecordMutation((List) modded, vars);
+        } else if (modded instanceof String && vars != null && !vars.isEmpty() && !isStringSafe((String)modded, vars, SAFE_ENVIRONMENT_VARIABLES)) {
+            this.isUnmodifiedBySanitization = false;
+            return NotStoredReason.MASKED_VALUE;
+        }
+
+        if (modded != tempVal) {
+            // Sanitization stripped out some values, so we need to record that and return modified version
+            this.isUnmodifiedBySanitization = false;
+            return modded;
+        } else {  // Any mutation was just from exploding step/uninstantiated describable, and we can just use the original
+            return o;
+        }
+    }
+
+    /**
+     * Does sanitization of a step's params, returning the Describable arguments map,
      * as from {@link org.jenkinsci.plugins.workflow.steps.StepDescriptor#defineArguments(Step)}
+     * Returns original input if not modified.
+     * Due to restrictions in {@link org.jenkinsci.plugins.structs.describable.DescribableModel#instantiate(Map)}
+     *  we do not need to sanitize keys, which can only parameters to a {@link Describable}
      * @param stepArguments Arguments provided when creating the step
      * @param variables Environment variables to remove
      * @return Arguments map, with oversized values and values containing bound credentials stripped out.
      */
     @Nonnull
-    public static Map<String,Object> sanitizedStepArguments(@Nonnull Map<String, Object> stepArguments, @CheckForNull EnvVars variables) {
+    Map<String,Object> sanitizeMapAndRecordMutation(@Nonnull Map<String, Object> stepArguments, @CheckForNull EnvVars variables) {
```

Comment:
The name `stepArguments` is misleading, because these could be arguments to some nested `Describable`.